{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import glob\n",
    "import urllib\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from io import StringIO\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import AlignIO\n",
    "from Bio.Align.Applications import MuscleCommandline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup path \n",
    "pd.set_option('display.max_rows', 10000)\n",
    "os.chdir('/Users/rootqz/Desktop/ReyLab/project/BSH/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Build HMM using BSH gene cluster from Foley et al. 2023 Nat. Microbiol. paper\n",
    "\n",
    "# load bsh gene files\n",
    "cluster_seq = pd.read_csv('data/Foley23_cluster_seq.csv', sep = ',')\n",
    "cluster_meta = pd.read_csv('data/Foley23_cluster_meta.csv', sep = ',')\n",
    "cloned_seq = pd.read_csv('data/Foley23_cloned_bsh.csv', sep = ',')\n",
    "\n",
    "# write out BSH cluster aa sequence fasta file\n",
    "with open('data/Foley23_cluster_seq.fasta', 'w') as out:\n",
    "    for i in range(cluster_seq.shape[0]):\n",
    "        print('>{}'.format(cluster_seq.cluster_name[i]), file=out)\n",
    "        print('{}'.format(cluster_seq.aa_seq[i]), file=out)\n",
    "\n",
    "# mulyiple alignment via MUSCLE\n",
    "subprocess.call('muscle3.8.31_i86darwin64 -in data/Foley23_cluster_seq.fasta \\\n",
    "                                          -out data/Foley23_cluster_seq.afa', stdout=True, shell=True)\n",
    "\n",
    "# convert to stockholm format\n",
    "hmm_bsh = SeqIO.parse('data/Foley23_cluster_seq.afa', 'fasta')\n",
    "SeqIO.write(hmm_bsh, 'data/Foley23_cluster_seq.sto', 'stockholm')\n",
    "\n",
    "# build hmm profile\n",
    "subprocess.call('hmmbuild data/Foley23_cluster_seq.hmm \\\n",
    "                          data/Foley23_cluster_seq.sto', stdout=True, shell=True)\n",
    "\n",
    "# build blast reference\n",
    "subprocess.call('makeblastdb -in data/blast_ref/Foley23_cluster_seq.fasta -dbtype prot', stdout=True, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Download CDS from bacteria strain genome\n",
    "\n",
    "# load strain list\n",
    "strain_list = pd.read_csv('data/strain_list.csv', sep = ',', index_col=0, na_values=None)\n",
    "\n",
    "# ncbi genomes list \n",
    "gcf_list = [x.split('/')[-2] for x in strain_list.ncbi_ftp.to_list() if isinstance(x, str)]\n",
    "\n",
    "# fetch gene sequence\n",
    "os.chdir('/Users/rootqz/Desktop/ReyLab/project/BSH/data/genome_faa/')\n",
    "\n",
    "for i in range(strain_list.shape[0]):\n",
    "    \n",
    "    # filter out NA value, i.e., no RefSeq genome from NCBI\n",
    "    if isinstance(strain_list.ncbi_ftp.to_list()[i], str):\n",
    "        GCF_ID = strain_list.ncbi_ftp.to_list()[i].split('/')[-2]\n",
    "        #print(GCF_ID)\n",
    "        \n",
    "        # download faa from RefSeq\n",
    "        subprocess.call('wget {}{}_protein.faa.gz'.format(strain_list.ncbi_ftp.to_list()[i], GCF_ID), stdout=True, shell=True)\n",
    "        # unzip\n",
    "        subprocess.call('gunzip {}_protein.faa.gz'.format(GCF_ID), stdout=True, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### BLASTp to BSH gene cluster \n",
    "\n",
    "for gcf in gcf_list:\n",
    "    \n",
    "    query_file = 'data/genome_faa/{}_protein.faa'.format(gcf)\n",
    "    out_file = 'data/blastp_out/{}.blastp'.format(gcf)\n",
    "    \n",
    "    # BLASTp\n",
    "    subprocess.call('blastp -query {} -db data/blast_ref/Foley23_cluster_seq.fasta -out {} -evalue 1e-5 -outfmt \"6 qseqid qlen sseqid slen pident nident ppos positive length mismatch gapopen qstart qend sstart send evalue bitscore\"'.format(query_file, out_file), stdout=True, shell=True)\n",
    "\n",
    "# concat BLASTp results into single df\n",
    "blastp_merge= pd.DataFrame()\n",
    "\n",
    "for gcf in gcf_list:\n",
    "    blastp_out_file = 'data/blastp_out/{}.blastp'.format(gcf)\n",
    "    if os.stat(blastp_out_file).st_size > 0:\n",
    "        blastp_out = pd.read_csv(blastp_out_file, sep = '\\t', header=None)\n",
    "        blastp_out.columns = ['qseqid', 'qlen', 'sseqid', 'slen', \n",
    "                              'pident', 'nident', 'ppos', 'positive', 'length', 'mismatch', 'gapopen', \n",
    "                              'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore']\n",
    "        \n",
    "        # add new column to store genome (GCF ID) info\n",
    "        blastp_out['gcfID'] = gcf\n",
    "        \n",
    "        if blastp_merge.shape[0] == 0:\n",
    "            blastp_merge = blastp_out\n",
    "        else:\n",
    "            blastp_merge = pd.concat([blastp_merge, blastp_out], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### HMM\n",
    "for gcf in gcf_list:\n",
    "    \n",
    "    query_file = 'data/genome_faa/{}_protein.faa'.format(gcf)\n",
    "    out_file = 'data/hmm_out/{}.hmm.out'.format(gcf)\n",
    "    \n",
    "    # BLASTp\n",
    "    subprocess.call('hmmsearch --tblout {} data/Foley23_cluster_seq.hmm {}'.format(out_file, query_file), stdout=True, shell=True)  \n",
    "\n",
    "# concat HMM results into single df\n",
    "hmm_merge = pd.DataFrame()\n",
    "\n",
    "for gcf in gcf_list:\n",
    "    hmm_out_file = 'data/hmm_out/{}.hmm.out'.format(gcf)\n",
    "    \n",
    "    # test if no HMM hit, thus hmm out file contains only comment lines\n",
    "    dummy_pd = pd.read_csv(hmm_out_file, sep='\\t', header=None)\n",
    "    \n",
    "    if dummy_pd.shape[0] > 13:\n",
    "        hmm_out = pd.read_csv(hmm_out_file, delim_whitespace=True, comment='#', header=None, usecols=range(18))\n",
    "        \n",
    "        hmm_out.columns = ['tname', 'tacc', 'qname', 'qacc', \n",
    "                           'full_evalue', 'full_score', 'full_bias', \n",
    "                           'domain_evalue', 'domain_score', 'domain_bias', \n",
    "                           'exp', 'reg', 'clu', 'ov', 'env', 'dom', 'rep', 'inc']\n",
    "        \n",
    "        # add new column to store genome (GCF ID) info\n",
    "        hmm_out['gcfID'] = gcf\n",
    "        \n",
    "        if hmm_merge.shape[0] == 0:\n",
    "            hmm_merge = hmm_out\n",
    "        else:\n",
    "            hmm_merge = pd.concat([hmm_merge, hmm_out], axis=0)\n",
    "\n",
    "# get gene by BLASTp criteria: gene_length > 300 and < 400, piden > 30\n",
    "blast_filter = blastp_merge[(blastp_merge['qlen'] > 300) & (blastp_merge['qlen'] < 400) & (blastp_merge['pident'] > 25)]\n",
    "blast_filter_nr = list(set(blast_filter['qseqid'].to_list()))\n",
    "\n",
    "# filter hit\n",
    "hmm_filter = hmm_merge[(hmm_merge['full_score'] > 100) & (hmm_merge['tname'].isin(blast_filter_nr))]\n",
    "\n",
    "bsh_hits = []\n",
    "bsh_seqID = []\n",
    "\n",
    "for i in range(hmm_filter.shape[0]):\n",
    "    gcf = hmm_filter['gcfID'].to_list()[i]\n",
    "    gene = hmm_filter['tname'].to_list()[i]\n",
    "\n",
    "    if gene not in bsh_seqID:\n",
    "        faa_file = 'data/genome_faa/{}_protein.faa'.format(gcf)\n",
    "        faa_seqs = SeqIO.parse(faa_file, 'fasta')\n",
    "        \n",
    "        # append all record to a list\n",
    "        bsh_hits = bsh_hits + [seq for seq in faa_seqs if seq.id == gene]\n",
    "        bsh_seqID = bsh_seqID+[gene]\n",
    "\n",
    "SeqIO.write(bsh_hits, 'data/BSH_hmm_hits.fasta', 'fasta')\n",
    "\n",
    "# merge gene acc with bacteria name\n",
    "strain_list_merge = strain_list[strain_list['ncbi_ftp'].notna()]\n",
    "strain_list_merge['gcfID'] = strain_list_merge['ncbi_ftp'].str.split('/', expand=True)[9]\n",
    "\n",
    "\n",
    "motif_merge = pd.merge(hmm_filter[['tname', 'full_evalue', 'full_score', 'gcfID']], \n",
    "                       strain_list_merge[['gcfID', 'name', 'phylum']], \n",
    "                       how='inner', \n",
    "                       on='gcfID')\n",
    "\n",
    "strain_list_merge[['gcfID', 'name', 'phylum']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Build HMM using PVA gene from Oâ€™Flaherty et al. 2018 mSphere paper\n",
    "\n",
    "# mulyiple alignment via MUSCLE\n",
    "subprocess.call('muscle3.8.31_i86darwin64 -in data/PVA_ref.fasta \\\n",
    "                                          -out data/PVA_ref.afa', stdout=True, shell=True)\n",
    "\n",
    "# convert to stockholm format\n",
    "hmm_bsh = SeqIO.parse('data/PVA_ref.afa', 'fasta')\n",
    "SeqIO.write(hmm_bsh, 'data/PVA_ref.sto', 'stockholm')\n",
    "\n",
    "# build hmm profile\n",
    "subprocess.call('hmmbuild data/PVA_ref.hmm \\\n",
    "                          data/PVA_ref.sto', stdout=True, shell=True)\n",
    "\n",
    "\n",
    "# hmm search\n",
    "subprocess.call('hmmsearch --tblout data/PVA_BSHhit.hmm.out data/PVA_ref.hmm data/BSH_hmm_hits.fasta', stdout=True, shell=True)  \n",
    "\n",
    "# convert hmm out table\n",
    "hmm_out = pd.read_csv('data/PVA_BSHhit.hmm.out', delim_whitespace=True, comment='#', header=None, usecols=range(18))\n",
    "\n",
    "hmm_out.columns = ['tname', 'tacc', 'qname', 'qacc', \n",
    "                   'full_evalue', 'full_score', 'full_bias', \n",
    "                   'domain_evalue', 'domain_score', 'domain_bias', \n",
    "                   'exp', 'reg', 'clu', 'ov', 'env', 'dom', 'rep', 'inc']\n",
    "\n",
    "hmm_out.to_csv('data/PVA_BSHhit_hmm.tsv', sep = '\\t', index=False)\n",
    "\n",
    "# hmm search of PVA ref seq\n",
    "subprocess.call('hmmsearch --tblout data/PVA_RefGene.hmm.out data/PVA_ref.hmm data/PVA_ref.fasta', stdout=True, shell=True)  \n",
    "\n",
    "# convert hmm out table\n",
    "hmm_out = pd.read_csv('data/PVA_RefGene.hmm.out', delim_whitespace=True, comment='#', header=None, usecols=range(18))\n",
    "\n",
    "hmm_out.columns = ['tname', 'tacc', 'qname', 'qacc', \n",
    "                   'full_evalue', 'full_score', 'full_bias', \n",
    "                   'domain_evalue', 'domain_score', 'domain_bias', \n",
    "                   'exp', 'reg', 'clu', 'ov', 'env', 'dom', 'rep', 'inc']\n",
    "\n",
    "hmm_out.to_csv('data/PVA_RefGene_hmm.tsv', sep = '\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
